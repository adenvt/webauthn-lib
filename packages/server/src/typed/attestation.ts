export type AttestationFormat = 'fido-u2f' | 'android-key' | 'android-safetynet' | 'packed' | 'tpm' | 'none'

export interface AttestationObject<T> {
  readonly fmt: AttestationFormat;
  readonly authData: Buffer;
  readonly attStmt: T;
}

export interface FidoU2FStatement {
  readonly sig: Buffer;
  readonly x5c: Buffer[];
}

export interface AndroidSafetyNetStatement {
  readonly ver: string;
  readonly response: Buffer;
}

export interface PackedSelfStatement {
  readonly alg: number;
  readonly sig: Buffer;
}

export interface PackedX5CStatement extends PackedSelfStatement {
  readonly x5c: Buffer[];
}

export interface PackedECDAAStatement extends PackedSelfStatement {
  readonly ecdaaKeyId: Buffer;
}

export interface AndroidKeyStatement {
  readonly alg: number;
  readonly sig: Buffer;
  readonly x5c: Buffer[];
}

export interface TPMStatement {
  readonly ver: string;
  readonly sig: Buffer;
  readonly certInfo: Buffer;
  readonly pubArea: Buffer;
  readonly alg: number;
}

export interface TPMX5CStatement extends TPMStatement {
  readonly x5c: Buffer[];
}

export interface TPMECDAAStatement extends TPMStatement {
  readonly ecdaaKeyId: Buffer;
}

export type NoneStatement = {}

export type FidoU2FAttestationObject = AttestationObject<FidoU2FStatement>

export type AndroidSafetyNetAttestationObject = AttestationObject<AndroidSafetyNetStatement>

export type PackedSelfAttestationObject = AttestationObject<PackedSelfStatement>

export type PackedX5CAttestationObject = AttestationObject<PackedX5CStatement>

export type PackedECDAAAttestationObject = AttestationObject<PackedECDAAStatement>

export type AndroidKeyAttestationObject = AttestationObject<AndroidKeyStatement>

export type TPMX5CAttestationObject = AttestationObject<TPMX5CStatement>

export type TPMECDAAAttestationObject = AttestationObject<TPMECDAAStatement>

export type NoneAttestationObject = AttestationObject<NoneStatement>
