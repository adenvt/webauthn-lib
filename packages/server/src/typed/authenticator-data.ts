export interface AuthenticatorDataFlag {
  readonly UP: boolean;
  readonly UV: boolean;
  readonly AT: boolean;
  readonly ED: boolean;
}

export interface AttestedCredentialData {
  readonly aaguid: string;
  readonly credentialIdLength: number;
  readonly credentialId: Buffer;
  readonly credentialPublicKey: Buffer;
}

export interface AuthenticatorData {
  readonly rpIdHash: Buffer;
  readonly flags: AuthenticatorDataFlag;
  readonly signCount: number;
  readonly attestedCredentialData?: AttestedCredentialData;
  readonly extensions?: Buffer;
}
