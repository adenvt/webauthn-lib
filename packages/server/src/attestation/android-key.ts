import base64url from 'base64url'
import parseAuthData from '../parser/parse-auth-data'
import { UserPubKey } from '../typed/webauthn'
import { AndroidKeyAttestationObject } from '../typed/attestation'
import { Certificate, Extension } from '@fidm/x509'
import { ASN1hasTag, ASN1parse } from '../utils/asn1'
import {
  hash,
  verifySignature,
  DERtoPEM,
  verifyCertPath,
  COSEtoPEM,
} from '../utils/crypto'

export default function verifyAndroidKeyAttestation (attsObject: AndroidKeyAttestationObject, clientDataJSON: string): UserPubKey {
  const authData = parseAuthData(attsObject.authData)

  /* istanbul ignore if */
  if (!authData.attestedCredentialData)
    throw new Error('authData must property attestedCredentialData')

  const credentialId = authData.attestedCredentialData.credentialId
  const certificates = (attsObject.attStmt.x5c)
    .map((cert) => DERtoPEM(cert))

  /**
  * Verify Certificate Chain
  */
  /* istanbul ignore if */
  if (verifyCertPath(certificates) !== true)
    throw new Error('Failed to verify Certificate Path')

  /**
   * Verify that sig is a valid signature over the concatenation of
   * authenticatorData and clientDataHash using the attestation public key in
   * attestnCert with the algorithm specified in alg.
   */
  const PEMCertificate = certificates[0]
  const signature      = attsObject.attStmt.sig
  const clientDataHash = hash('SHA256', base64url.toBuffer(clientDataJSON))
  const signatureBase  = Buffer.concat([attsObject.authData, clientDataHash])
  const verified       = verifySignature(signature, signatureBase, PEMCertificate)

  /* istanbul ignore if */
  if (!verified)
    throw new Error('Failed to verify Attestation Signature')

  /**
   * Verify that the public key in the first certificate in x5c matches
   * the credentialPublicKey in the attestedCredentialData in authenticatorData.
   */
  const attestationCert = Certificate.fromPEM(Buffer.from(PEMCertificate))
  const authPubKey      = COSEtoPEM(authData.attestedCredentialData.credentialPublicKey)
  const certPubKey      = attestationCert.publicKey.toPEM()

  /* istanbul ignore if */
  if (authPubKey.trim() !== certPubKey.trim())
    throw new Error('Certificate public key does not match public key in authData')

  const extension     = attestationCert.getExtension('1.3.6.1.4.1.11129.2.1.17') as Extension
  const extensionData = ASN1parse(extension.value)
  const extensionHash = extensionData.data[4].data

  /**
   * Verify that the attestationChallenge field in the attestation certificate
   * extension data is identical to clientDataHash.
   */
  /* istanbul ignore if */
  if (!Buffer.isBuffer(extensionHash) || !extensionHash.equals(clientDataHash))
    throw new Error('Certificate attestation challenge is not set to the clientData hash!')

  /**
   * The AuthorizationList.allApplications field is not present on either
   * authorization list (softwareEnforced nor teeEnforced),
   * since PublicKeyCredential MUST be scoped to the RP ID.
   */
  const softwareEnforced = extensionData.data[6].data
  const teeEnforced      = extensionData.data[7].data

  /* istanbul ignore if */
  if (ASN1hasTag(softwareEnforced, 600) || ASN1hasTag(teeEnforced, 600))
    throw new Error('TEE or Software autherisation cannot contains "allApplication" flag')

  return {
    fmt         : 'android-key',
    counter     : authData.signCount,
    credentialId: base64url.encode(credentialId),
    publicKey   : authPubKey,
  }
}
