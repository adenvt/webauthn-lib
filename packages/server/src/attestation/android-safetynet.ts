import base64url from 'base64url'
import { AndroidSafetyNetAttestationObject } from '../typed/attestation'
import { Certificate } from '@fidm/x509'
import { UserPubKey } from '../typed/webauthn'
import parseAuthData from '../parser/parse-auth-data'
import {
  hash,
  verifySignature,
  Base64toPEM,
  verifyCertPath,
  COSEtoPEM,
} from '../utils/crypto'

interface SafetyNetPayload {
  readonly timestampMs: number;
  readonly nonce: string;
  readonly apkPackageName: string;
  readonly apkCertificateDigestSha256: string[];
  readonly ctsProfileMatch: boolean;
  readonly basicIntegrity: boolean;
}

const ROOT_CERT = `
-----BEGIN CERTIFICATE-----
MIIDujCCAqKgAwIBAgILBAAAAAABD4Ym5g0wDQYJKoZIhvcNAQEFBQAwTDEgMB4G
A1UECxMXR2xvYmFsU2lnbiBSb290IENBIC0gUjIxEzARBgNVBAoTCkdsb2JhbFNp
Z24xEzARBgNVBAMTCkdsb2JhbFNpZ24wHhcNMDYxMjE1MDgwMDAwWhcNMjExMjE1
MDgwMDAwWjBMMSAwHgYDVQQLExdHbG9iYWxTaWduIFJvb3QgQ0EgLSBSMjETMBEG
A1UEChMKR2xvYmFsU2lnbjETMBEGA1UEAxMKR2xvYmFsU2lnbjCCASIwDQYJKoZI
hvcNAQEBBQADggEPADCCAQoCggEBAKbPJA6+Lm8omUVCxKs+IVSbC9N/hHD6ErPL
v4dfxn+G07IwXNb9rfF73OX4YJYJkhD10FPe+3t+c4isUoh7SqbKSaZeqKeMWhG8
eoLrvozps6yWJQeXSpkqBy+0Hne/ig+1AnwblrjFuTosvNYSuetZfeLQBoZfXklq
tTleiDTsvHgMCJiEbKjNS7SgfQx5TfC4LcshytVsW33hoCmEofnTlEnLJGKRILzd
C9XZzPnqJworc5HGnRusyMvo4KD0L5CLTfuwNhv2GXqF4G3yYROIXJ/gkwpRl4pa
zq+r1feqCapgvdzZX99yqWATXgAByUr6P6TqBwMhAo6CygPCm48CAwEAAaOBnDCB
mTAOBgNVHQ8BAf8EBAMCAQYwDwYDVR0TAQH/BAUwAwEB/zAdBgNVHQ4EFgQUm+IH
V2ccHsBqBt5ZtJot39wZhi4wNgYDVR0fBC8wLTAroCmgJ4YlaHR0cDovL2NybC5n
bG9iYWxzaWduLm5ldC9yb290LXIyLmNybDAfBgNVHSMEGDAWgBSb4gdXZxwewGoG
3lm0mi3f3BmGLjANBgkqhkiG9w0BAQUFAAOCAQEAmYFThxxol4aR7OBKuEQLq4Gs
J0/WwbgcQ3izDJr86iw8bmEbTUsp9Z8FHSbBuOmDAGJFtqkIk7mpM0sYmsL4h4hO
291xNBrBVNpGP+DTKqttVCL1OmLNIG+6KYnX3ZHu01yiPqFbQfXf5WRDLenVOavS
ot+3i9DAgBkcRcAtjOj4LaR0VknFBbVPFd5uRHg5h6h+u/N5GJG79G+dwfCMNYxd
AfvDbbnvRG15RjF+Cv6pgsH/76tuIMRQyV+dTZsXjAzlAcmgQWpzU/qlULRuJQ/7
TBj0/VLZjmmx6BEP3ojY+x1J96relc8geMJgEtslQIxq/H5COEBkEveegeGTLg==
-----END CERTIFICATE-----
`

export default function verifyAndroidSafetyNetAttestation (attsObject: AndroidSafetyNetAttestationObject, clientDataJSON: string): UserPubKey {
  const authData = parseAuthData(attsObject.authData)

  /* istanbul ignore if */
  if (!authData.attestedCredentialData)
    throw new Error('authData must property attestedCredentialData')

  const response     = attsObject.attStmt.response.toString('utf-8')
  const jwsResponse  = response.split('.')
  const header       = JSON.parse(base64url.decode(jwsResponse[0]))
  const payload      = JSON.parse(base64url.decode(jwsResponse[1])) as SafetyNetPayload
  const signature    = base64url.toBuffer(jwsResponse[2])
  const certificates = (header.x5c as string[])
    .map((cert) => Base64toPEM(cert))
    .concat([ROOT_CERT])

  /**
  * Verify Certificate Chain
  */
  /* istanbul ignore if */
  if (verifyCertPath(certificates) !== true)
    throw new Error('Failed to verify Certificate Path')

  /**
   * Verify that the nonce in the response is identical to the Base64 encoding
   * of the SHA-256 hash of the concatenation of authenticatorData and clientDataHash.
   */
  const clientDataHash = hash('SHA256', base64url.toBuffer(clientDataJSON))
  const nonceBuffer    = hash('SHA256', Buffer.concat([attsObject.authData, clientDataHash]))
  const nonce          = nonceBuffer.toString('base64')

  /* istanbul ignore if */
  if (payload.nonce !== nonce)
    throw new Error('Failed to verify "nonce" value')

  /**
   * Verify that attestationCert is issued to the hostname "attest.android.com"
   */
  const PEMCertificate  = certificates[0]
  const attestationCert = Certificate.fromPEM(Buffer.from(PEMCertificate))

  /* istanbul ignore if */
  if (attestationCert.subject.commonName !== 'attest.android.com')
    throw new Error('Certificate Subject CN must be "attest.android.com"')

  /**
   * Verify that the ctsProfileMatch attribute in the payload of response is true.
   */
  /* istanbul ignore if */
  if (!payload.ctsProfileMatch)
    throw new Error('Payload.ctsProfileMatch must be true')

  const signatureBase = Buffer.from(`${jwsResponse[0]}.${jwsResponse[1]}`)
  const verified      = verifySignature(signature, signatureBase, PEMCertificate)

  /* istanbul ignore if */
  if (!verified)
    throw new Error('Failed to verify Attestation Signature')

  const credentialId = authData.attestedCredentialData.credentialId
  const publicKey    = COSEtoPEM(authData.attestedCredentialData.credentialPublicKey)

  return {
    fmt         : 'android-safetynet',
    publicKey   : publicKey,
    counter     : authData.signCount,
    credentialId: base64url.encode(credentialId),
  }
}
