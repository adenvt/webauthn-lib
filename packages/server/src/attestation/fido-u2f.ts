import { FidoU2FAttestationObject } from '../typed/attestation'
import parseAuthData from '../parser/parse-auth-data'
import base64url from 'base64url'
import { UserPubKey } from '../typed/webauthn'
import COSEtoJWK from 'cose-to-jwk'
import {
  hash,
  verifySignature,
  DERtoPEM,
  JWKtoPEM,
} from '../utils/crypto'

export default function verifyFidoU2FAttestation (attsObject: FidoU2FAttestationObject, clientDataJSON: string): UserPubKey {
  const authData = parseAuthData(attsObject.authData)

  /* istanbul ignore if */
  if (!authData.attestedCredentialData)
    throw new Error('authData must have property attestedCredentialData')

  const rpIdHash       = authData.rpIdHash
  const credentialId   = authData.attestedCredentialData.credentialId
  const clientDataHash = hash('SHA256', base64url.toBuffer(clientDataJSON))
  const pubKeyJWK      = COSEtoJWK(authData.attestedCredentialData.credentialPublicKey)

  /**
   * If certificate public key is not an Elliptic Curve (EC) public key over the
   * P-256 curve, terminate this algorithm and return an appropriate error.
   */
  /* istanbul ignore if */
  if (pubKeyJWK.kty !== 'EC' || pubKeyJWK.crv !== 'P-256')
    throw new Error('Public Key must be Elliptic Curve (EC) over P-256 curve')

  /* istanbul ignore if */
  if (!pubKeyJWK.x || !pubKeyJWK.y)
    throw new Error('Public Key must be contains "x" and "y"')

  const publicKeyU2F = Buffer.concat([
    Buffer.from([0x04]),
    Buffer.from(pubKeyJWK.x, 'base64'),
    Buffer.from(pubKeyJWK.y, 'base64'),
  ])

  /**
   * Let verificationData be the concatenation of
   * (0x00 || rpIdHash || clientDataHash || credentialId || publicKeyU2F)
   */
  const signatureBase = Buffer.concat([
    Buffer.from([0x00]), // Reserved Bytes
    rpIdHash,
    clientDataHash,
    credentialId,
    publicKeyU2F,
  ])

  const PEMCertificate = DERtoPEM(attsObject.attStmt.x5c[0])
  const signature      = attsObject.attStmt.sig
  const verified       = verifySignature(signature, signatureBase, PEMCertificate)

  /* istanbul ignore if */
  if (!verified)
    throw new Error('Failed to verify Attestation Signature')

  return {
    fmt         : 'fido-u2f',
    counter     : authData.signCount,
    credentialId: base64url.encode(credentialId),
    publicKey   : JWKtoPEM(pubKeyJWK),
  }
}
