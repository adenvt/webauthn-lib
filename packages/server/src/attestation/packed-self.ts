import base64url from 'base64url'
import parseAuthData from '../parser/parse-auth-data'
import { UserPubKey } from '../typed/webauthn'
import { PackedSelfAttestationObject } from '../typed/attestation'
import {
  hash,
  verifySignature,
  COSEtoPEM,
} from '../utils/crypto'

export default function verifyPackedSelf (attsObject: PackedSelfAttestationObject, clientDataJSON: string): UserPubKey {
  const authData = parseAuthData(attsObject.authData)

  /* istanbul ignore if */
  if (!authData.attestedCredentialData)
    throw new Error('authData must property attestedCredentialData')

  const credentialId   = authData.attestedCredentialData.credentialId
  const PEMCertificate = COSEtoPEM(authData.attestedCredentialData.credentialPublicKey)

  const alg       = attsObject.attStmt.alg
  const signature = attsObject.attStmt.sig

  /* istanbul ignore if */
  if (alg !== -7)
    throw new Error('Unsupported Attestation Signature algorithm, only support ES256')

  const clientDataHash = hash('SHA256', base64url.toBuffer(clientDataJSON))
  const signatureBase  = Buffer.concat([attsObject.authData, clientDataHash])
  const verified       = verifySignature(signature, signatureBase, PEMCertificate)

  /* istanbul ignore if */
  if (!verified)
    throw new Error('Failed to verify Attestation Signature')

  return {
    fmt         : 'packed',
    publicKey   : PEMCertificate,
    counter     : authData.signCount,
    credentialId: base64url.encode(credentialId),
  }
}
