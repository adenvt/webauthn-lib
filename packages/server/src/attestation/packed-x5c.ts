import base64url from 'base64url'
import { Certificate } from '@fidm/x509'
import parseAuthData from '../parser/parse-auth-data'
import isISO31661Alpha2 from '../utils/iso-3166'
import { UserPubKey } from '../typed/webauthn'
import { PackedX5CAttestationObject } from '../typed/attestation'
import {
  hash,
  verifySignature,
  DERtoPEM,
  COSEtoPEM,
} from '../utils/crypto'

export default function verifyPackedX5C (attsObject: PackedX5CAttestationObject, clientDataJSON: string): UserPubKey {
  const authData = parseAuthData(attsObject.authData)

  /* istanbul ignore if */
  if (!authData.attestedCredentialData)
    throw new Error('authData must property attestedCredentialData')

  const PEMCertificate  = DERtoPEM(attsObject.attStmt.x5c[0])
  const attestationCert = Certificate.fromPEM(Buffer.from(PEMCertificate))
  const subject         = attestationCert.subject

  /**
   * Getting requirements from https://www.w3.org/TR/webauthn/#packed-attestation
   * Version MUST be set to 3 (which is indicated by an ASN.1 INTEGER with value 2).
   */
  /* istanbul ignore if */
  if (attestationCert.version !== 3)
    throw new Error('Version MUST be set to 3 (which is indicated by an ASN.1 INTEGER with value 2)')

  /**
   * Subject-C
   * ISO 3166 code specifying the country where the Authenticator vendor is incorporated (PrintableString)
   */
  /* istanbul ignore if */
  if (!isISO31661Alpha2(subject.countryName))
    throw new Error('Invalid Subject Country, must be valid ISO 3166')

  /**
   * Subject-O
   * Legal name of the Authenticator vendor (UTF8String)
   */
  /* istanbul ignore if */
  if (!subject.organizationName)
    throw new Error('Invalid Subject Organization Name')

  /**
   * Subject-OU
   * Literal string "Authenticator Attestation" (UTF8String)
   */
  /* istanbul ignore if */
  if (subject.organizationalUnitName !== 'Authenticator Attestation')
    throw new Error('Invalid Subject Organizational Unit Name, must be "Authenticator Attestation"')

  /**
   * Subject-CN
   * A UTF8String of the vendorâ€™s choosing
   */
  /* istanbul ignore if */
  if (!subject.commonName)
    throw new Error('Invalid Subject Common Name')

  /**
   * The Basic Constraints extension MUST have the CA component set to false
   */
  /* istanbul ignore if */
  if (attestationCert.isCA !== false)
    throw new Error('Basic Constraints extension MUST have the CA component set to false')

  /**
   * If the related attestation root certificate is used for multiple authenticator models,
   * the Extension OID 1.3.6.1.4.1.45724.1.1.4 (id-fido-gen-ce-aaguid) MUST be present,
   * containing the AAGUID as a 16-byte OCTET STRING.
   */
  const aaguidExtension = attestationCert.getExtension('1.3.6.1.4.1.45724.1.1.4')
  /* istanbul ignore if */
  if (aaguidExtension !== null) {
    // AAGUID MUST be wrapped in two OCTET STRINGS
    if (aaguidExtension.value.slice(2).toString('hex') !== authData.attestedCredentialData.aaguid)
      throw new Error('Invalid Extension "id-fido-gen-ce-aaguid" value')

    // The extension MUST NOT be marked as critical.
    if (aaguidExtension.critical !== false)
      throw new Error('Extension "id-fido-gen-ce-aaguid" must not be marked as critical')
  }

  /**
   * Verify that sig is a valid signature over the concatenation of
   * authenticatorData and clientDataHash using the attestation public key in
   * attestnCert with the algorithm specified in alg.
   */
  const signature      = attsObject.attStmt.sig
  const clientDataHash = hash('SHA256', base64url.toBuffer(clientDataJSON))
  const signatureBase  = Buffer.concat([attsObject.authData, clientDataHash])
  const verified       = verifySignature(signature, signatureBase, PEMCertificate)

  /* istanbul ignore if */
  if (!verified)
    throw new Error('Failed to verify Attestation Signature')

  const credentialId = authData.attestedCredentialData.credentialId
  const publicKey    = COSEtoPEM(authData.attestedCredentialData.credentialPublicKey)

  return {
    fmt         : 'packed',
    publicKey   : publicKey,
    counter     : authData.signCount,
    credentialId: base64url.encode(credentialId),
  }
}
