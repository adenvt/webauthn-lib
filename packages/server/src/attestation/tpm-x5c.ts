import { NoneAttestationObject } from '../typed/attestation'
import { UserPubKey } from '../typed/webauthn'
import parseAuthData from '../parser/parse-auth-data'
import base64url from 'base64url'
import { COSEtoPEM } from '../utils/crypto'

export default function verifyTPMX5C (attsObject: NoneAttestationObject, _clientDataJSON: string): UserPubKey {
  const authData = parseAuthData(attsObject.authData)

  /* istanbul ignore if */
  if (!authData.attestedCredentialData)
    throw new Error('authData must property attestedCredentialData')

  const credentialId = authData.attestedCredentialData.credentialId
  const publicKey    = COSEtoPEM(authData.attestedCredentialData.credentialPublicKey)

  return {
    fmt         : 'none',
    publicKey   : publicKey,
    counter     : authData.signCount,
    credentialId: base64url.encode(credentialId),
  }
}
