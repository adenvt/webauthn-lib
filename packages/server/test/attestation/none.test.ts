import attestationResponse from '../__sample__/none.json'
import verifyNoneAttestation from '../../src/attestation/none'
import parseAttestationObject from '../../src/parser/parse-attestation-object'
import { NoneAttestationObject } from '../../src/typed/attestation'

test('Verify FidoU2F', () => {
  const attsObject = parseAttestationObject(attestationResponse.response.attestationObject) as NoneAttestationObject
  const result     = verifyNoneAttestation(attsObject, attestationResponse.response.clientDataJSON)

  expect(result).toHaveProperty('fmt')
  expect(result).toHaveProperty('counter')
  expect(result).toHaveProperty('publicKey')
  expect(result).toHaveProperty('credentialId')

  expect(result).toMatchSnapshot()
})
