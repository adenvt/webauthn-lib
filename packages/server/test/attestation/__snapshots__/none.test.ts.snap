// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify FidoU2F 1`] = `
Object {
  "counter": 1,
  "credentialId": "Y4s-fgVIXbT7NqvHa1sLMd7T7ErhZn_qF4wa2Yc-Uj4",
  "fmt": "none",
  "publicKey": "-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAErtJAmExf4K22LwBG1rr/HNnvIkbV
9bGGzouNSzRFbQIpCGhlsNp6jzkvjgVUNIdzYmeUEiAl7qvsr8xo06NM8A==
-----END PUBLIC KEY-----",
}
`;
